src/workflow.ts

```diff
@@ -205,51 +205,40 @@ export class WorkflowContextImpl extends DBOSContextImpl implements WorkflowCont
+  async recordOutput<R>(client: UserDatabaseClient, funcID: number, output: R): Promise<string> {
+    try {
+      const serialOutput = JSON.stringify(output);
+      const rows = await this.#dbosExec.userDatabase.queryWithClient<transaction_outputs>(client, "INSERT INTO dbos.transaction_outputs (workflow_uuid, function_id, output, txn_id, txn_snapshot, created_at) VALUES ($1, $2, $3, (select pg_current_xact_id_if_assigned()::text), (select pg_current_snapshot()::text), $4) RETURNING txn_id;", this.workflowUUID, funcID, serialOutput, Date.now());
+      return rows[0].txn_id;
+    } catch (error) {
+      if (this.#dbosExec.userDatabase.isKeyConflictError(error)) {
+        // Serialization and primary key conflict (Postgres).
+        throw new DBOSWorkflowConflictUUIDError(this.workflowUUID);
+      } else {
+        throw error;
+      }
     }
-    this.resultBuffer.set(funcID, guardOutput);
-  }
-
-  /**
-   * Write a guarded operation's output to the database.
-   */
-  async recordGuardedOutput<R>(client: UserDatabaseClient, funcID: number, output: R): Promise<string> {
-    const serialOutput = JSON.stringify(output);
-    const rows = await this.#dbosExec.userDatabase.queryWithClient<transaction_outputs>(client, "UPDATE dbos.transaction_outputs SET output=$1, txn_id=(select pg_current_xact_id_if_assigned()::text) WHERE workflow_uuid=$2 AND function_id=$3 RETURNING txn_id;", serialOutput, this.workflowUUID, funcID);
-    return rows[0].txn_id;  // Must have a transaction ID because we inserted the guard before.
-  }
-
-  /**
-   * Write an unguarded operation's output to the database.
-   */
-  async recordUnguardedOutput<R>(client: UserDatabaseClient, funcID: number, output: R): Promise<string> {
-    const serialOutput = JSON.stringify(output);
-    const rows = await this.#dbosExec.userDatabase.queryWithClient<transaction_outputs>(client, "INSERT INTO dbos.transaction_outputs (workflow_uuid, function_id, output, txn_id, txn_snapshot, created_at) VALUES ($1, $2, $3, (select pg_current_xact_id_if_assigned()::text), (select pg_current_snapshot()::text), $4) RETURNING txn_id;", this.workflowUUID, funcID, serialOutput, Date.now());
-    return rows[0].txn_id;
   }
 
   /**
-   * Record an error in a guarded operation to the database.
+   * Record an error in an operation to the database.
    */
-  async recordGuardedError(client: UserDatabaseClient, funcID: number, err: Error) {
-    const serialErr = JSON.stringify(serializeError(err));
-    await this.#dbosExec.userDatabase.queryWithClient(client, "UPDATE dbos.transaction_outputs SET error=$1 WHERE workflow_uuid=$2 AND function_id=$3;", serialErr, this.workflowUUID, funcID);
-  }
-
-  /**
-   * Record an error in an unguarded operation to the database.
-   */
-    async recordUnguardedError(client: UserDatabaseClient, funcID: number, err: Error): Promise<string> {
+  async recordError(client: UserDatabaseClient, funcID: number, err: Error): Promise<string> {
+    try {
```

Comment:
Since the original tx failed, should we be passing in txn_id or txn_snapshot rather than getting the current values?

Reply:
That's a good point. Will fix that.